core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -413,85 +468,90 @@ import com.datastax.driver.core.utils.UUIDs
-            state = QueryIdle
-            missingLookup = missingLookup.map(m => {
+            log.info("[{}] {}: Still looking for missing. {}. Waiting for next poll.", stageUuid, session.tag, stageState)
+            updateStageState(_.copy(missingLookup = stageState.missingLookup.map(m => {
               val newBucket = TimeBucket(m.maxOffset, bucketSize)
               m.copy(
                 bucket = newBucket,
-                queryPrevious = !newBucket.within(m.previousOffset)
-              )
-            })
+                queryPrevious = !newBucket.within(m.previousOffset))
+            })))
             // a current query doesn't have a poll we schedule
-            if (refreshInterval.isEmpty)
+            if (!isLiveQuery())
               scheduleOnce(QueryPoll, settings.refreshInterval)
           }
-        } else if (currTimeBucket.inPast) {
+        } else if (stageState.shouldMoveBucket()) {
           nextTimeBucket()
-          log.debug("Moving to next bucket: {}", currTimeBucket)
+          log.debug("[{}] Moving to next bucket: {}", stageUuid, stageState.currentTimeBucket)
           query()
         } else {
-          if (toOffset.isDefined) {
-            log.debug("At today, ending.")
+          if (toOffset.contains(stageState.toOffset)) {
+            log.debug("[{}] Current query finished and has passed the eventual consistency delay, ending.", stageUuid)
             completeStage()
           } else {
-            state = QueryIdle
-            log.debug("Nothing todo, waiting for next poll")
+            log.debug("[{}] Nothing todo, waiting for next poll", stageUuid)
+            if (!isLiveQuery()) {
+              // As this isn't a live query schedule a poll for now
+              log.debug("[{}] Scheduling poll for eventual consistency delay: {}", stageUuid, settings.eventsByTagEventualConsistency)
+              scheduleOnce(QueryPoll, settings.eventsByTagEventualConsistency)
+            }
           }
         }
       }
 
       private def stopLookingForMissing(m: LookingForMissing, buffered: List[UUIDRow]): Unit = {
-        state = if (buffered.isEmpty) {
+        updateQueryState(if (buffered.isEmpty) {
           QueryIdle
         } else {
           BufferedEvents(buffered.sortBy(_.tagPidSequenceNr))
-        }
-        log.info("No more missing events. Sending buffered events. {}", state)
-        currentOffset = m.maxOffset
-        currTimeBucket = TimeBucket(m.maxOffset, bucketSize)
-        tagPidSequenceNrs += (m.persistenceId -> ((m.maxSequenceNr, m.maxOffset)))
-        missingLookup = None
+        })
+        log.info("[{}] No more missing events. Sending buffered events. {}", stageUuid, stageState.state)
+        updateStageState(_.copy(fromOffset = m.maxOffset))
+        updateStageState(state =>
+          state.copy(missingLookup = None).tagPidSequenceNumberUpdate(m.persistenceId, (m.maxSequenceNr, m.maxOffset)))
```

Comment:
Hmm, why the two transitions instead of just one?

Reply:
The problem was between the screen and the chair